AsyncDisplayKit/ASImageNode+AnimatedImage.mm

```diff
@@ -0,0 +1,226 @@
+//
+//  ASImageNode+AnimatedImage.m
+//  AsyncDisplayKit
+//
+//  Created by Garrett Moon on 3/22/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASImageNode+AnimatedImage.h"
+
+#import "ASAssert.h"
+#import "ASImageProtocols.h"
+#import "ASDisplayNode+Subclasses.h"
+#import "ASDisplayNodeExtras.h"
+#import "ASEqualityHelpers.h"
+#import "ASDisplayNode+FrameworkPrivate.h"
+#import "ASImageNode+AnimatedImagePrivate.h"
+#import "ASInternalHelpers.h"
+#import "ASWeakProxy.h"
+
+@implementation ASImageNode (AnimatedImage)
+
+#pragma mark - GIF support
+
+- (void)setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  if (!ASObjectIsEqual(_animatedImage, animatedImage)) {
+    _animatedImage = animatedImage;
+  }
+  if (animatedImage != nil) {
+    __weak ASImageNode *weakSelf = self;
+    if ([animatedImage respondsToSelector:@selector(setCoverImageReadyCallback:)]) {
+      animatedImage.coverImageReadyCallback = ^(UIImage *coverImage) {
+        [weakSelf coverImageCompleted:coverImage];
+      };
+    }
+    
+    animatedImage.playbackReadyCallback = ^{
+      [weakSelf animatedImageFileReady];
+    };
+  }
+}
+
+- (id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  return _animatedImage;
+}
+
+- (void)setAnimatedImagePaused:(BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
```

Comment:
Do we need a separate lock for this?  Reduced contention is one thing, but not likely an issue here, and multiple locks can get pretty complicated.